---
title: "celltype_example"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{celltype_example}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
### Load libraries
library(CompassR)
library(dplyr)
library(Seurat)
library(Gviz)
```

Here, we will explore the celltype specific analysis of the gene HEY2 in cardiomyocytes of atrium and ventricle. 
```{r}
### set up the gene and celltype
gene = "HEY2"
celltype = "Cardiomyocyte"
```

We will first query the expression and linkage data for the gene HEY2 in cardiomyocytes from CompassDB. 
```{r}
expr_vec = query_exprssion("hg38", gene, celltype = TRUE)
link_res = query_linkage("hg38", gene, celltype = TRUE)

link_df = link_res[["linkage"]]
sample_df = link_res[["samples"]]

### filter the table, extract features use for plotting
sample_df = sample_df[sample_df$bio_source %in% c("Heart right ventricle", "Left cardiac atrium"), ]
sample_df = sample_df[, c("sample_id", "bio_source")]
colnames(sample_df) = c("Sample", "Group")

### prepare linkage data
link_df = link_df[link_df$sample %in% sample_df$Sample & link_df$celltype == celltype, ]
# link_df[, c("sample", "gene", "peak", "zscore", "source", "celltype")]

### prepare sample information
sample_df = sample_df[sample_df$Sample %in% link_df$sample, ]
sample_df = sample_df[order(sample_df$Group), ]

### prepare gene expression
expr_df = data.frame(t(expr_vec)[paste0(sample_df$Sample, ".", sample_df$Sample, ".RNA.", celltype), ])
colnames(expr_df) = c("Gene")
rownames(expr_df) = sapply(rownames(expr_df), function(x) {unlist(strsplit(x, "\\."))[1]})
expr_df$Sample = rownames(expr_df)

link_df$sample_id = link_df$sample
nrow(link_df)
```

We will then plot the expression of the gene in the samples and the linkage data. 
```{r}
scheme <- getScheme()
scheme$GdObject$cex.axis = 1
addScheme(scheme, "myScheme")
options(Gviz.scheme="myScheme")

sample_df$Group = c(rep("Ventricle", 2), rep("Atrium", 2))
# pdf("Heatmap_celltype_cadiomyocytes_HEY2.pdf", width = 10, height = 6)
genome_track_map(link_df, sample_df, gene, expr_df, assembly = "hg38", legend.position="left", t = -30, b = 30)
# dev.off()
```

TFBS analysis is performed on the CREs linked to the gene HEY2 in the cardiomyocytes.
```{r}
### TFBS Analysis
track_plot = plot_genome_track(link_df, "HEY2", "hg38", sample_df$Sample)
peaks = track_plot[[length(track_plot)]]
peaks$name = as.character(peaks$name)
peak_order = make_peak_group(link_df, peaks, sample_df)

### Find CREs linked to HEY2 in ventricle but not in atrium
ht_peak_order = peak_order[peak_order[, "Ventricle"] >= 0.5 & peak_order[, "Atrium"]<0.5, ]
ht_peaks = rownames(ht_peak_order)

peak_vec = sub(":", "-", as.character(peaks))
ht_peaks = peak_vec[as.numeric(ht_peaks)]

tf = tf_binding("hg38", paste0(ht_peaks, collapse = "_"))
plot_giggle(tf)
# ggsave("Barplot_HEY2_Ventricle_specific_TFs.pdf", p, width = 5, height = 5)
```

TFBS analysis results can also be visualized as a wordcloud.
```{r}
library(ggwordcloud)

ggplot(tf[1:50, ], aes(label = Factor, size = combo_score)) +
  geom_text_wordcloud() +
  theme_minimal()
# p
# ggsave("Wordcloud_HEY2_Ventricle_specific_TFs.pdf", p, width = 6, height = 5)

```

